### PR TITLE
Variable Render Props

### PR DESCRIPTION
**Pull request by kadenbarlow**
The SDK currently renders hyperlink variables as straight text and doesn't do anything to support actually rendering a link. There is no one size fit all to replacing all link variables with a single component that will work in all cases. This PR introduces a generalized way to interpolate any variable type although it was made in particular to handle replacing hyperlink variables with actual links in the text. 

```typescript
<Ditto
  componentId="component-with-hyperlink"
  richText
  variables={{
    link: (variableData: DittoVariableData) => {
      if (DittoVariableTypeGuards.isHyperlink(variableData)) {
        return <a href={variableData.url}>{variableData.text}</a>
      }
    }
  }}
/>
```

### CODE REVIEW

File: src/lib/context.ts [hunk 1]
```diff
@@ -157,6 +158,24 @@ export const SourceDetector = {
...
+    return "__type" in variableData && variableData.__type === "map";
+  },
+  isNumber: function (variableData: VariableData): variableData is VariableNumber {
+    return "__type" in variableData && variableData.__type === "number";
+  },
+  isString: function (variableData: VariableData): variableData is VariableString {
+    return "__type" in variableData && variableData.__type === "string";
+  },
+};
+
```
Comment (azjgard) [inline src/lib/context.ts:178]: It's worth noting that in the event the user is using `flat` or `icu` formats, we don't have access to any of the variable metadata, and that's something that we need to somehow account for

Comment (kadenbarlow) [reply src/lib/context.ts:178]: Good point, I've updated the `VariableRenderPropType` to make the variable metadata optional. When using the flat or icu formats that data would be null and previous to this change we'd just return whatever input is passed into the variable. See the updated tests for examples of what that would look like in code.

File: src/hooks/useDittoSingleText.ts [hunk 2]
```diff
@@ -63,7 +63,8 @@ export const useDittoSingleText = (props: useDittoSingleTextProps): string | nul
   }
 
   if (SourceDetector.isStructured(data) || SourceDetector.isFlat(data)) {
-    return interpolateVariableText(data[textId], variables || {}, count, false).text;
+    return interpolateVariableText(data[textId], variables || {}, count, Boolean(props.richText))
```
Comment (kadenbarlow) [inline src/hooks/useDittoSingleText.ts:66]: We weren't respecting the richText prop for useDittoSingleText hook

File: tests/app/README.md
Comment (azjgard) [inline tests/app/README.md:1]: Maybe we should remove this file?

**Review by azjgard — APPROVED** [review]
Looks great!

**Pull request merged**